### PR TITLE
feat: Convert line filter expressions to a set of bloom tests

### PR DESCRIPTION
**Pull request by salvacorts**
**What this PR does / why we need it**:
This PR introduces:
- A new interface `BloomTest` with two test methods that take a bloom as an argument and returns a boolean. True if the test passes with the given bloom.
- A new function `FiltersToBloomTest` that given a set of `LineFilterExpr`, it returns a `BloomTest`.

~The logic in this PR is not yet used anywhere since we need to do some refactoring to be able to use it. These refactorings will be done on a set of follow-up PRs:~
~1. Pass `syntax.LineFilterExpr` instead of `syntax.LineFilter` with `logproto.FilterChunkRefRequest`~
~2. Create BloomTest from the line filters passed along through `logproto.FilterChunkRefRequest` and modify the fused querier to use the resulting `BloomTest`.~
3. Map regex line filters to `BloomTest` when possible. The logic would be pretty similar to what's done in the `simplify` function at [pkg/logql/log/filter.go](https://github.com/grafana/loki/blob/60edfe0006b0adf6041b7ac437b8c47d66d1ead6/pkg/logql/log/filter.go#L470-L496). We can probably extract the simplification logic into [pkg/util/regex.go].(https://github.com/grafana/loki/blob/5a08a6bcb9ecc816328cb7859bbdcc7cf9d004c5/pkg/util/regex.go)

**Special notes for your reviewer**:
- As explained in the bullets above, regexes are not yet tested. all regex filter expressions are mapped to Noop tests which always return true.

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: pkg/storage/bloom/v1/bloom_tester.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+import (
+	"github.com/prometheus/prometheus/model/labels"
+
+	"github.com/grafana/loki/pkg/logql/syntax"
+	"github.com/grafana/loki/pkg/storage/bloom/v1/filter"
+)
+
+type BloomTest interface {
+	Test(bloom filter.Checker) bool
```
Comment (chaudum) [inline pkg/storage/bloom/v1/bloom_tester.go:11]: nit: `Test()` does not give any clue about what the returned `true` or `false` means.

`BloomTest` is a [predicate](https://en.wikipedia.org/wiki/Predicate_%28mathematical_logic%29) of `Request`.

```go
type MatchesBloom interface {
    ...
}
```

or

```go
struct BloomTest interface {
    Matches(bloom filter.Checker) bool
    ...
}
```

File: pkg/storage/bloom/v1/bloom_tester.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	case labels.MatchRegexp, labels.MatchNotRegexp:
+		// TODO(salvacorts): Simplify regex similarly to how it's done at pkg/logql/log/filter.go (`simplify` function)
+		// 					 Ideally we want to extract the simplify logic into pkg/util/regex.go
+		return NoopTest
+	default:
+		return NoopTest
+	}
+}
+
+type noopTest struct{}
```
Comment (chaudum) [inline pkg/storage/bloom/v1/bloom_tester.go:70]: nit like above: this could be named `matchAll`

File: pkg/storage/bloom/v1/bloom_tester_test.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+func extractLineFilters(expr syntax.Expr) (lineFilters []syntax.LineFilterExpr) {
+	visitor := &syntax.DepthFirstTraversal{
+		VisitLineFilterFn: func(v syntax.RootVisitor, e *syntax.LineFilterExpr) {
+			lineFilters = append(lineFilters, *e)
+		},
+	}
+	expr.Accept(visitor)
+	return lineFilters
+}
```
Comment (chaudum) [inline pkg/storage/bloom/v1/bloom_tester_test.go:137]: I think you can use the function from the `syntax` package, right?

File: pkg/storage/bloom/v1/bloom_tester_test.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+type fakeNgramBuilder struct{}
```
Comment (chaudum) [inline pkg/storage/bloom/v1/bloom_tester_test.go:139]: this is actually not a fake ngram builder, but a very simple tokenizer, right? 😀

File: pkg/storage/bloom/v1/tokenizer.go [hunk 5]
```diff
@@ -42,8 +42,8 @@ func NewNGramTokenizer(n, skip int) *NGramTokenizer {
 
 // The Token iterator uses shared buffers for performance. The []byte returned by At()
 // is not safe for use after subsequent calls to Next()
-func (t *NGramTokenizer) Tokens(line string) NGramTokenIter {
-	return NGramTokenIter{
+func (t *NGramTokenizer) Tokens(line string) Iterator[[]byte] {
```
Comment (chaudum) [inline pkg/storage/bloom/v1/tokenizer.go:45]: Can you add a comment that this function implements the new `NGramBuilder` interface?

File: pkg/storage/bloom/v1/tokenizer.go [hunk 6]
```diff
@@ -97,17 +97,17 @@ type PrefixedTokenIter struct {
 	buf       []byte
 	prefixLen int
 
-	NGramTokenIter
+	Iterator[[]byte]
```
Comment (chaudum) [inline pkg/storage/bloom/v1/tokenizer.go:100]: I wouldn't make this field exported.

**Review by chaudum — COMMENTED** [review]
This is looking really good!

File: pkg/storage/bloom/v1/tokenizer.go (hunk 6 above)
Comment (salvacorts) [reply pkg/storage/bloom/v1/tokenizer.go:100]: Done

File: pkg/storage/bloom/v1/tokenizer.go (hunk 5 above)
Comment (salvacorts) [reply pkg/storage/bloom/v1/tokenizer.go:45]: Done

File: pkg/storage/bloom/v1/bloom_tester_test.go (hunk 3 above)
Comment (salvacorts) [reply pkg/storage/bloom/v1/bloom_tester_test.go:137]: Indeed, leftover from before rebasing

File: pkg/storage/bloom/v1/bloom_tester.go (hunk 2 above)
Comment (salvacorts) [reply pkg/storage/bloom/v1/bloom_tester.go:70]: Done

File: pkg/storage/bloom/v1/tokenizer.go (hunk 6 above)
Comment (salvacorts) [reply pkg/storage/bloom/v1/tokenizer.go:100]: it's indirectly  exported. We need have that way so we don't need to reimplement Next, Err...

File: pkg/storage/bloom/v1/bloom_tester_test.go (hunk 4 above)
Comment (salvacorts) [reply pkg/storage/bloom/v1/bloom_tester_test.go:129]: Yes, but w/o much utility other than testing

File: pkg/storage/bloom/v1/tokenizer.go (hunk 6 above)
Comment (salvacorts) [reply pkg/storage/bloom/v1/tokenizer.go:101]: I don't have a stong opinion on this, i'm fine with making this private and implementing Err and Next as wrappers. But I think this is more idiomatic

**Review by owen-d — APPROVED** [review]

**Pull request merged**